File: py/zbc/encryption.py
# SPDX-License-Identifier: MIT
# Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci
"""Sealed transaction messages (spec/signing.md section 8): 'ZBE1' || libsodium sealed box to the recipient's converted key.
X25519, HSalsa20, XSalsa20 and Poly1305 on the standard library (integers and hashlib), so the package keeps its no-dependency promise."""
import hashlib
import os
import struct
from typing import Optional

P = 2 ** 255 - 19

def x25519(k: bytes, u: bytes) -> bytes:
    """RFC 7748 X25519(scalar bytes, u-coordinate bytes) -> 32 bytes."""
    kk = bytearray(k); kk[0] &= 248; kk[31] &= 127; kk[31] |= 64
    s = int.from_bytes(kk, "little"); x1 = int.from_bytes(u, "little") & ((1 << 255) - 1)
    x2, z2, x3, z3, swap = 1, 0, x1, 1, 0
    for t in range(254, -1, -1):
        kt = (s >> t) & 1
        swap ^= kt
        if swap: x2, x3 = x3, x2; z2, z3 = z3, z2
        swap = kt
        a = (x2 + z2) % P; aa = a * a % P; b = (x2 - z2) % P; bb = b * b % P; e = (aa - bb) % P
        c = (x3 + z3) % P; d = (x3 - z3) % P; da = d * a % P; cb = c * b % P
        x3 = pow((da + cb) % P, 2, P); z3 = x1 * pow((da - cb) % P, 2, P) % P
        x2 = aa * bb % P; z2 = e * ((aa + 121665 * e) % P) % P
    if swap: x2, x3 = x3, x2; z2, z3 = z3, z2
    return (x2 * pow(z2, P - 2, P) % P).to_bytes(32, "little")

def x25519_base(k: bytes) -> bytes: return x25519(k, (9).to_bytes(32, "little"))

def ed25519_pk_to_x25519(pk: bytes) -> bytes:
    """u = (1 + y) / (1 - y) mod p."""
    y = int.from_bytes(pk, "little") & ((1 << 255) - 1)
    return ((1 + y) * pow(1 - y, P - 2, P) % P).to_bytes(32, "little")

def ed25519_seed_to_x25519(seed: bytes) -> bytes:
    h = bytearray(hashlib.sha512(seed).digest()[:32]); h[0] &= 248; h[31] &= 127; h[31] |= 64
    return bytes(h)

_SIGMA = b"expand 32-byte k"
def _rotl(v, c): return ((v << c) & 0xFFFFFFFF) | (v >> (32 - c))
def _salsa_rounds(x):
    for _ in range(10):
        for (a, b, c, d) in ((0, 4, 8, 12), (5, 9, 13, 1), (10, 14, 2, 6), (15, 3, 7, 11), (0, 1, 2, 3), (5, 6, 7, 4), (10, 11, 8, 9), (15, 12, 13, 14)):
            x[b] ^= _rotl((x[a] + x[d]) & 0xFFFFFFFF, 7); x[c] ^= _rotl((x[b] + x[a]) & 0xFFFFFFFF, 9)
            x[d] ^= _rotl((x[c] + x[b]) & 0xFFFFFFFF, 13); x[a] ^= _rotl((x[d] + x[c]) & 0xFFFFFFFF, 18)
    return x

def hsalsa20(key: bytes, inp: bytes) -> bytes:
    """HSalsa20(key 32, input 16) -> 32 bytes (no feed-forward; words 0,5,10,15,6,7,8,9)."""
    c = struct.unpack("<4I", _SIGMA); k = struct.unpack("<8I", key); n = struct.unpack("<4I", inp)
    x = [c[0], k[0], k[1], k[2], k[3], c[1], n[0], n[1], n[2], n[3], c[2], k[4], k[5], k[6], k[7], c[3]]
    x = _salsa_rounds(x)
    return struct.pack("<8I", x[0], x[5], x[10], x[15], x[6], x[7], x[8], x[9])

def salsa20_block(key: bytes, nonce8: bytes, counter: int) -> bytes:
    c = struct.unpack("<4I", _SIGMA); k = struct.unpack("<8I", key); n = struct.unpack("<2I", nonce8)
    x0 = [c[0], k[0], k[1], k[2], k[3], c[1], n[0], n[1], counter & 0xFFFFFFFF, (counter >> 32) & 0xFFFFFFFF, c[2], k[4], k[5], k[6], k[7], c[3]]
    x = _salsa_rounds(list(x0))
    return struct.pack("<16I", *[(x[i] + x0[i]) & 0xFFFFFFFF for i in range(16)])

def xsalsa20_stream(key: bytes, nonce24: bytes, length: int) -> bytes:
    sub = hsalsa20(key, nonce24[:16]); out = b""; i = 0
    while len(out) < length: out += salsa20_block(sub, nonce24[16:], i); i += 1
    return out[:length]

def poly1305(key32: bytes, msg: bytes) -> bytes:
    r = int.from_bytes(key32[:16], "little") & 0x0ffffffc0ffffffc0ffffffc0fffffff; s = int.from_bytes(key32[16:], "little")
    p = (1 << 130) - 5; acc = 0
    for i in range(0, len(msg), 16):
        blk = msg[i:i + 16]; n = int.from_bytes(blk, "little") + (1 << (8 * len(blk)))
        acc = (acc + n) * r % p
    return ((acc + s) & ((1 << 128) - 1)).to_bytes(16, "little")

def secretbox(key: bytes, nonce: bytes, plaintext: bytes) -> bytes:
    """crypto_secretbox_easy: tag (16) || ciphertext."""
    stream = xsalsa20_stream(key, nonce, 32 + len(plaintext))
    c = bytes(a ^ b for a, b in zip(plaintext, stream[32:]))
    return poly1305(stream[:32], c) + c

def secretbox_open(key: bytes, nonce: bytes, boxed: bytes):
    if len(boxed) < 16: return None
    stream = xsalsa20_stream(key, nonce, 32 + len(boxed) - 16)
    if poly1305(stream[:32], boxed[16:]) != boxed[:16]: return None
    return bytes(a ^ b for a, b in zip(boxed[16:], stream[32:]))

def box_key(sk: bytes, pk: bytes) -> bytes: return hsalsa20(x25519(sk, pk), bytes(16))

def blake2b_24(data: bytes) -> bytes: return hashlib.blake2b(data, digest_size=24).digest()

SEALED_MAGIC = b"ZBE1"       # the 4-byte marker in front of a sealed message field
SEALED_OVERHEAD = 52         # marker 4 + ephemeral key 32 + tag 16


def is_sealed(field: bytes) -> bool:
    """True when the field starts with the marker (it may still fail to open)."""
    return len(field) >= 4 and field[:4] == SEALED_MAGIC


def seal(plaintext: bytes, recipient_public_key: bytes, ephemeral_secret_key: Optional[bytes] = None) -> bytes:
    """Seal `plaintext` to the recipient's 32-byte Ed25519 public key. The ephemeral secret key is random unless given (tests)."""
    if len(recipient_public_key) != 32:
        raise ValueError("recipient public key must be 32 bytes")
    esk = ephemeral_secret_key if ephemeral_secret_key is not None else os.urandom(32)
    if len(esk) != 32:
        raise ValueError("ephemeral secret key must be 32 bytes")
    rpk = ed25519_pk_to_x25519(recipient_public_key); epk = x25519_base(esk)
    return SEALED_MAGIC + epk + secretbox(box_key(esk, rpk), blake2b_24(epk + rpk), plaintext)


def open_sealed(field: bytes, recipient_seed: bytes) -> Optional[bytes]:
    """The plaintext of a sealed field opened with the recipient's 32-byte seed, or None when it is not a sealed message or the key does not open it."""
    if not is_sealed(field) or len(field) < SEALED_OVERHEAD or len(recipient_seed) != 32:
        return None
    sk = ed25519_seed_to_x25519(recipient_seed); pk = x25519_base(sk); epk = field[4:36]
    return secretbox_open(box_key(sk, epk), blake2b_24(epk + pk), field[36:])

File: go/zbc/encryption.go
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci

package zbc

import (
	"bytes"
	"crypto/ecdh"
	"crypto/rand"
	"crypto/sha512"
	"encoding/binary"
	"errors"
	"math/big"
)

// Sealed transaction messages (spec/signing.md section 8): "ZBE1" || libsodium sealed box to the
// recipient's Ed25519 key converted to X25519. X25519 comes from crypto/ecdh; HSalsa20, XSalsa20
// and Poly1305 are here, since the standard library has neither Salsa20 nor Poly1305.

// SealedMagic is the 4-byte marker in front of a sealed message field.
var SealedMagic = []byte("ZBE1")

// SealedOverhead is how much longer a sealed field is than its plaintext: marker 4 + ephemeral key 32 + tag 16.
const SealedOverhead = 52

var curveP = new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), 255), big.NewInt(19))

func leToBig(b []byte) *big.Int {
	r := make([]byte, len(b))
	for i := range b {
		r[len(b)-1-i] = b[i]
	}
	return new(big.Int).SetBytes(r)
}

func bigToLe(v *big.Int, n int) []byte {
	out := make([]byte, n)
	b := v.Bytes()
	for i := range b {
		out[i] = b[len(b)-1-i]
	}
	return out
}

// Ed25519PublicKeyToX25519 converts an Ed25519 public key to the X25519 public key of the same
// point: u = (1 + y) / (1 - y) mod p.
func Ed25519PublicKeyToX25519(pk []byte) ([]byte, error) {
	if len(pk) != 32 {
		return nil, errors.New("public key must be 32 bytes")
	}
	y := leToBig(pk)
	y.SetBit(y, 255, 0)
	one := big.NewInt(1)
	num := new(big.Int).Add(one, y)
	den := new(big.Int).Sub(one, y)
	den.Mod(den, curveP)
	if den.Sign() == 0 {
		return nil, errors.New("public key has no X25519 form")
	}
	u := num.Mul(num, den.ModInverse(den, curveP))
	return bigToLe(u.Mod(u, curveP), 32), nil
}

// Ed25519SeedToX25519 is the X25519 secret key of an Ed25519 seed: the clamped first half of SHA-512(seed).
func Ed25519SeedToX25519(seed []byte) []byte {
	h := sha512.Sum512(seed)
	sk := h[:32]
	sk[0] &= 248
	sk[31] &= 127
	sk[31] |= 64
	return sk
}

// X25519Base is the X25519 public key of a scalar: X25519(scalar, 9).
func X25519Base(scalar []byte) ([]byte, error) {
	k, err := ecdh.X25519().NewPrivateKey(scalar)
	if err != nil {
		return nil, err
	}
	return k.PublicKey().Bytes(), nil
}

func x25519(scalar, u []byte) ([]byte, error) {
	k, err := ecdh.X25519().NewPrivateKey(scalar)
	if err != nil {
		return nil, err
	}
	p, err := ecdh.X25519().NewPublicKey(u)
	if err != nil {
		return nil, err
	}
	return k.ECDH(p)
}

var salsaSigma = [4]uint32{0x61707865, 0x3320646e, 0x79622d32, 0x6b206574} // "expand 32-byte k"

func rotl(v uint32, c uint) uint32 { return v<<c | v>>(32-c) }

func salsaRounds(x *[16]uint32) {
	for i := 0; i < 10; i++ {
		for _, q := range [8][4]int{{0, 4, 8, 12}, {5, 9, 13, 1}, {10, 14, 2, 6}, {15, 3, 7, 11}, {0, 1, 2, 3}, {5, 6, 7, 4}, {10, 11, 8, 9}, {15, 12, 13, 14}} {
			a, b, c, d := q[0], q[1], q[2], q[3]
			x[b] ^= rotl(x[a]+x[d], 7)
			x[c] ^= rotl(x[b]+x[a], 9)
			x[d] ^= rotl(x[c]+x[b], 13)
			x[a] ^= rotl(x[d]+x[c], 18)
		}
	}
}

func salsaState(key []byte, x *[16]uint32) {
	x[0], x[5], x[10], x[15] = salsaSigma[0], salsaSigma[1], salsaSigma[2], salsaSigma[3]
	for i := 0; i < 4; i++ {
		x[1+i] = binary.LittleEndian.Uint32(key[4*i:])
		x[11+i] = binary.LittleEndian.Uint32(key[16+4*i:])
	}
}

// HSalsa20 is HSalsa20(key 32, input 16) -> 32 bytes.
func HSalsa20(key, input []byte) []byte {
	var x [16]uint32
	salsaState(key, &x)
	for i := 0; i < 4; i++ {
		x[6+i] = binary.LittleEndian.Uint32(input[4*i:])
	}
	salsaRounds(&x)
	out := make([]byte, 32)
	for i, w := range [8]int{0, 5, 10, 15, 6, 7, 8, 9} {
		binary.LittleEndian.PutUint32(out[4*i:], x[w])
	}
	return out
}

func salsa20Block(key, nonce8 []byte, counter uint64) [64]byte {
	var x0 [16]uint32
	salsaState(key, &x0)
	x0[6] = binary.LittleEndian.Uint32(nonce8)
	x0[7] = binary.LittleEndian.Uint32(nonce8[4:])
	x0[8] = uint32(counter)
	x0[9] = uint32(counter >> 32)
	x := x0
	salsaRounds(&x)
	var out [64]byte
	for i := 0; i < 16; i++ {
		binary.LittleEndian.PutUint32(out[4*i:], x[i]+x0[i])
	}
	return out
}

// XSalsa20Stream is the XSalsa20 keystream of a 32-byte key and a 24-byte nonce.
func XSalsa20Stream(key, nonce24 []byte, length int) []byte {
	sub := HSalsa20(key, nonce24[:16])
	out := make([]byte, 0, length+64)
	for c := uint64(0); len(out) < length; c++ {
		blk := salsa20Block(sub, nonce24[16:24], c)
		out = append(out, blk[:]...)
	}
	return out[:length]
}

// Poly1305 is the one-time authenticator of RFC 8439 section 2.5.
func Poly1305(key32, msg []byte) []byte {
	r := leToBig(key32[:16])
	r.And(r, leToBig([]byte{0xff, 0xff, 0xff, 0x0f, 0xfc, 0xff, 0xff, 0x0f, 0xfc, 0xff, 0xff, 0x0f, 0xfc, 0xff, 0xff, 0x0f}))
	s := leToBig(key32[16:32])
	p := new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), 130), big.NewInt(5))
	acc := new(big.Int)
	for i := 0; i < len(msg); i += 16 {
		end := i + 16
		if end > len(msg) {
			end = len(msg)
		}
		n := leToBig(msg[i:end])
		n.SetBit(n, 8*(end-i), 1)
		acc.Add(acc, n)
		acc.Mul(acc, r)
		acc.Mod(acc, p)
	}
	acc.Add(acc, s)
	acc.And(acc, new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), 128), big.NewInt(1)))
	return bigToLe(acc, 16)
}

// Secretbox is crypto_secretbox_easy: tag (16) || ciphertext.
func Secretbox(key, nonce, plaintext []byte) []byte {
	stream := XSalsa20Stream(key, nonce, 32+len(plaintext))
	c := make([]byte, len(plaintext))
	for i := range plaintext {
		c[i] = plaintext[i] ^ stream[32+i]
	}
	return append(Poly1305(stream[:32], c), c...)
}

// SecretboxOpen is crypto_secretbox_open_easy: the plaintext, or false when the tag does not verify.
func SecretboxOpen(key, nonce, boxed []byte) ([]byte, bool) {
	if len(boxed) < 16 {
		return nil, false
	}
	c := boxed[16:]
	stream := XSalsa20Stream(key, nonce, 32+len(c))
	if !bytes.Equal(Poly1305(stream[:32], c), boxed[:16]) {
		return nil, false
	}
	out := make([]byte, len(c))
	for i := range c {
		out[i] = c[i] ^ stream[32+i]
	}
	return out, true
}

func boxKey(sk, pk []byte) ([]byte, error) {
	shared, err := x25519(sk, pk)
	if err != nil {
		return nil, err
	}
	return HSalsa20(shared, make([]byte, 16)), nil
}

func sealedNonce(ephemeralPk, recipientPkX []byte) []byte {
	return blake2b(append(append([]byte{}, ephemeralPk...), recipientPkX...), 24)
}

// IsSealed is true when the field starts with the marker (it may still fail to open).
func IsSealed(field []byte) bool { return len(field) >= 4 && bytes.Equal(field[:4], SealedMagic) }

// Seal seals plaintext to the recipient's 32-byte Ed25519 public key. The ephemeral secret key is
// drawn at random unless one is given (tests).
func Seal(plaintext, recipientPublicKey, ephemeralSecretKey []byte) ([]byte, error) {
	rpk, err := Ed25519PublicKeyToX25519(recipientPublicKey)
	if err != nil {
		return nil, err
	}
	esk := ephemeralSecretKey
	if esk == nil {
		esk = make([]byte, 32)
		if _, err := rand.Read(esk); err != nil {
			return nil, err
		}
	}
	epk, err := X25519Base(esk)
	if err != nil {
		return nil, err
	}
	key, err := boxKey(esk, rpk)
	if err != nil {
		return nil, err
	}
	out := append(append([]byte{}, SealedMagic...), epk...)
	return append(out, Secretbox(key, sealedNonce(epk, rpk), plaintext)...), nil
}

// OpenSealed opens a sealed field with the recipient's 32-byte Ed25519 seed: the plaintext, or
// false when it is not a sealed message or the key does not open it.
func OpenSealed(field, recipientSeed []byte) ([]byte, bool) {
	if !IsSealed(field) || len(field) < SealedOverhead || len(recipientSeed) != 32 {
		return nil, false
	}
	sk := Ed25519SeedToX25519(recipientSeed)
	pk, err := X25519Base(sk)
	if err != nil {
		return nil, false
	}
	epk := field[4:36]
	key, err := boxKey(sk, epk)
	if err != nil {
		return nil, false
	}
	return SecretboxOpen(key, sealedNonce(epk, pk), field[36:])
}
